src/services/staking_vaults.py

```diff
@@ -0,0 +1,659 @@
+                vault.vault_address,
+                vault.total_value_wei,
+                vault.fee,
+                vault.liability_shares,
+                vault.max_liability_shares,
+                vault.slashing_reserve,
+            ))
+
+        rebuild_merkle_tree = self.get_merkle_tree(tree_data)
+        root_hex = f'0x{rebuild_merkle_tree.root.hex()}'
+        return merkle_tree.tree[0] == root_hex and root_hex == expected_tree_root
+
+    @staticmethod
+    def calc_fee_value(value: Decimal, block_elapsed: int, core_apr_ratio: Decimal, fee_bp: int) -> Decimal:
+        return (
+            value
+            * Decimal(block_elapsed)
+            * core_apr_ratio
+            * Decimal(fee_bp)
+            / Decimal(BLOCKS_PER_YEAR * TOTAL_BASIS_POINTS)
+        )
+
+    @staticmethod
+    # pylint: disable=too-many-branches
+    def calc_liquidity_fee(
+        vault_address: str,
+        liability_shares: Shares,
+        liquidity_fee_bp: int,
+        events: defaultdict[str, list[VaultEventType]],
+        prev_block_number: BlockNumber,
+        current_block: BlockNumber,
+        pre_total_pooled_ether: Wei,
+        pre_total_shares: Shares,
+        core_apr_ratio: Decimal,
+    ) -> tuple[Decimal, Shares]:
+        """
+        Liquidity fee = Minted_stETH × Lido_Core_APR × Liquidity_fee_rate
+
+        We calculate the liquidity fee for the vault as a series of intervals
+        between minting, burning, and fee update events.
+
+        If there are no events, we just use `liability_shares` to compute minted stETH.
+
+        Burn: In the future, shares go down; backwards, they go up.
+        Mint: In the future, shares go up; backwards, they go down.
+
+        liability_shares (Y)
+                ↑
+                │
+                │
+                │
+                │          (shares were higher before burn)
+                │                 ┌──────────────
+                │                 │              │
+                │                 │              │
+                ┌─────────────────┘              │
+                │                 │              │  (shares decreased)
+                │                 │              │───────────│
+                │                 │              │           │
+                │                 │              │           │
+                └─────────────────┴──────────────┴───────────┴────────▶ block_number (X)
+
+                                 mintEvent      burnEvent  current_block
+
+                        ◄─────────────────────── processing backwards in time
+        """
+        vault_liquidity_fee = Decimal(0)
+        liquidity_fee = liquidity_fee_bp
+
+        if vault_address not in events:
+            blocks_elapsed = current_block - prev_block_number
+            minted_steth = get_steth_by_shares(liability_shares, pre_total_pooled_ether, pre_total_shares)
+            vault_liquidity_fee += StakingVaultsService.calc_fee_value(
```

Comment:
Let's return results right here
```suggestion
            return StakingVaultsService.calc_fee_value(...), liability_shares
```

Reply:
Plan for refactoring

Reply:
Added to #824